src/dataflow/operators/capture/event.rs

```diff
@@ -119,27 +119,32 @@ pub mod binary {
     use std::io::Write;
     use abomonation::Abomonation;
     use super::{Event, EventPusher, EventIterator};
+    use std::cell::RefCell;
 
     /// A wrapper for `W: Write` implementing `EventPusher<T, D>`.
     pub struct EventWriter<T, D, W: ::std::io::Write> {
-        stream: W,
+        buffer: RefCell<Vec<u8>>,
+        stream: RefCell<W>,
         phant: ::std::marker::PhantomData<(T,D)>,
     }
 
     impl<T, D, W: ::std::io::Write> EventWriter<T, D, W> {
         /// Allocates a new `EventWriter` wrapping a supplied writer.
         pub fn new(w: W) -> EventWriter<T, D, W> {
             EventWriter {
-                stream: w,
+                buffer: RefCell::new(vec![]),
+                stream: RefCell::new(w),
                 phant: ::std::marker::PhantomData,
             }
         }
     }
 
     impl<T: Abomonation, D: Abomonation, W: ::std::io::Write> EventPusher<T, D> for EventWriter<T, D, W> {
         fn push(&mut self, event: Event<T, D>) {
-            // TODO: `push` has no mechanism to report errors, so we `unwrap`.
-            unsafe { ::abomonation::encode(&event, &mut self.stream).unwrap(); }
+            let mut buffer = self.buffer.borrow_mut();
+            unsafe { ::abomonation::encode(&event, &mut *buffer).unwrap(); }
+            self.stream.borrow_mut().write_all(&buffer[..]).unwrap();
+            buffer.clear();
```

Comment:
Can you explain what's going on here? It looks like you have a second buffer into which you stage the serialized data, and then you copy it all out again. Is something else going on with the buffer that I'm missing? (( Also, the `RefCell` around the writer )). Perhaps this makes sense given constraints somewhere else!